android/jni/app-android.cpp

```diff
@@ -377,159 +656,223 @@ extern "C" void Java_org_ppsspp_ppsspp_NativeRenderer_displayRender(JNIEnv *env,
+	engine.postCommand = engine.jniEnv->GetMethodID(engine.jniEnv->GetObjectClass(engine.jniActivity), "postCommand", "(Ljava/lang/String;Ljava/lang/String;)V");
+	ILOG("Fetched method ID %p to postCommand...", engine.postCommand);
 
-	axis.axisId = JOYSTICK_AXIS_ACCELEROMETER_Y;
-	axis.value = y;
-	bool retvalY = NativeAxis(axis);
+	if (state->savedState != NULL) {
+		// We are starting with a previous saved state; restore from it.
+		engine.state = *(struct saved_state*)state->savedState;
+	}
 
-	axis.axisId = JOYSTICK_AXIS_ACCELEROMETER_Z;
-	axis.value = z;
-	bool retvalZ = NativeAxis(axis);
+	sampleRate = optimalSampleRate;
+	if (NativeQueryConfig("force44khz") != "0" || optimalSampleRate == 0) {
+		sampleRate = 44100;
+	}
+	if (optimalFramesPerBuffer > 0) {
+		framesPerBuffer = optimalFramesPerBuffer;
+	} else {
+		framesPerBuffer = 512;
+	}
 
-	return retvalX || retvalY || retvalZ;
-}
+	// Some devices have totally bonkers buffer sizes like 8192. They will have terrible latency anyway, so to avoid having to
+	// create extra smart buffering code, we'll just let their regular mixer deal with it, missing the fast path (as if they had one...)
+	if (framesPerBuffer > 512) {
+		framesPerBuffer = 512;
+		sampleRate = 44100;
+	}
 
-extern "C" void Java_org_ppsspp_ppsspp_NativeApp_sendMessage(JNIEnv *env, jclass, jstring message, jstring param) {
-	std::string msg = GetJavaString(env, message);
-	std::string prm = GetJavaString(env, param);
+	ILOG("NativeApp.audioInit() -- Using OpenSL audio! frames/buffer: %i   optimal sr: %i   actual sr: %i", optimalFramesPerBuffer, optimalSampleRate, sampleRate);
+	AndroidAudio_Init(&NativeMix, library_path, framesPerBuffer, sampleRate);
 
-	if (msg == "moga") {
-		mogaVersion = prm;
+	// Setup hardware scaling using getDesiredBackbufferSize
+	// ANativeWindow* window = ANativeWindow_fromSurface(env, surface);
+	// ANativeWindow_setBuffersGeometry(window, bufferWidth, bufferHeight, RGB_565);
+
+	// loop waiting for stuff to do.
+	while (1) {
+		// Read all pending events.
+		int ident;
+		int events;
+		struct android_poll_source* source;
+
+		// If not animating, we will block forever waiting for events.
+		// If animating, we loop until all events are read, then continue
+		// to draw the next frame of animation.
+		while ((ident = ALooper_pollAll(engine.animating ? 0 : -1, NULL, &events, (void**)&source)) >= 0) {
+
+			// Process this event.
+			if (source != NULL) {
+				source->process(state, source);
+			}
+
+			// If a sensor has data, process it immediately.
+			if (ident == LOOPER_ID_USER) {
+				if (engine.accelerometerSensor) {
+					ASensorEvent sensor_event;
+					while (ASensorEventQueue_getEvents(engine.sensorEventQueue, &sensor_event, 1) > 0) {
+						ProcessAccelerometer(sensor_event);
+					}
+				}
+			}
+
+			// Check if we are exiting.
+			if (state->destroyRequested != 0) {
+				ILOG("Destroy requested! Leaving.");
+				engine_term_display(&engine);
+				goto quit;
```

Comment:
I'm not one of those anti-goto people but, maybe a `engine.running` or `running` flag makes more sense, and just a break here?

Does `destroyRequested` get set by our code (not yet implemented yet, but based on a CMD) or directly by the Android library?

-[Unknown]